buildSrc/subprojects/parser/src/main/java/org/apache/calcite/buildtools/parser/DialectGenerate.java

```diff
@@ -120,10 +124,152 @@ public void unparseReservedKeywords(ExtractedData extractedData) {
+   *
+   * @param extractedData The object which keeps state of all of the extracted
+   *                      data
+   */
+  public void unparseNonReservedKeywords(ExtractedData extractedData) {
+    final ArrayList<Set<Keyword>> partitions = getPartitions(extractedData.nonReservedKeywords);
+    final int numPartitions = partitions.size();
+    final String declarationTemplate = "void NonReservedKeyword%dof"
+      + numPartitions + "():";
+    final StringBuilder bodyBuilder = new StringBuilder();
+    final List<String> functionCalls = new LinkedList<String>();
+    if (!partitions.isEmpty()) {
+      bodyBuilder.append("(\n");
+    }
+    for (int i = 0; i < numPartitions; i++) {
+      String declaration = String.format(declarationTemplate, i, numPartitions);
+      String functionName = getFunctionName(declaration);
+      extractedData.functions.put(functionName,
+          getPartitionFunction(declaration, partitions.get(i)));
+      functionCalls.add(functionName + "()\n");
+    }
+    bodyBuilder.append(String.join("| ", functionCalls));
+    if (!partitions.isEmpty()) {
+      bodyBuilder.append(")\n");
+    }
+    bodyBuilder.append("{ return unquotedIdentifier(); }\n");
+    String declaration = "void NonReservedKeyword():";
+    String function = buildAutoGeneratedFunction(declaration,
+        bodyBuilder.toString());
+    extractedData.functions.put("NonReservedKeyword", function);
+  }
+
+  /**
+   * Creates a string of the form:
+   *
+   * // Auto generated.
+   * <declaration>
+   * {
+   * }
+   * {
+   *         <KEYWORD1>
+   *     |   <KEYWORD2>
+   *     ...
+   * }
+   *
+   * Note: Indentation is added here just for clarity, actual function doesn't
+   * have indentation as it is difficult to format using String.join().
+   *
+   * @param declaration The function declaration
+   * @param keywords The keywords to be in the function body
+   *
+   * @return The formatted string
+   */
+  public String getPartitionFunction(String declaration, Set<Keyword> keywords) {
+    final List<String> tokens = new LinkedList<String>();
+    final StringBuilder bodyBuilder = new StringBuilder();
+    for (Keyword keyword : keywords) {
+      StringBuilder tokenBuilder = new StringBuilder();
+      tokenBuilder.append("<").append(keyword.keyword)
+        .append(">");
+      if (keyword.filePath == null) {
+        tokenBuilder.append(" // No file specified.");
+      } else {
+        tokenBuilder.append(" // From: " + keyword.filePath);
+      }
+      tokenBuilder.append("\n");
+      tokens.add(tokenBuilder.toString());
+    }
+    bodyBuilder.append(String.join("| ", tokens));
+    return buildAutoGeneratedFunction(declaration, bodyBuilder.toString());
+  }
+
+  /**
+   * Partitions {@code keywords} into sets of size at most 500.
+   *
+   * @param keywords The keywords to partition
+   *
+   * @return The partitioned keywords
+   */
+  private ArrayList<Set<Keyword>> getPartitions(Set<Keyword> keywords) {
```

Comment:
nit: Should we mark "keywords" as final to make it clear that we don't plan to modify it in this function?